File: src/root-query/parse-root-queries.test.ts
import { describe, expect, test } from "vitest";

import { parseRootQueries } from "./parse-root-queries.js";
import { ROOT_QUERY_KIND } from "./root-query-kind.js";

describe("parseRootQueries", () => {
  test("parses a single token", () => {
    const r = parseRootQueries("10:foo");
    expect(r.ok).toEqual(true);
    if (r.ok) {
      expect(r.queries).toHaveLength(1);
      expect(r.queries[0]).toMatchObject({
        kind: ROOT_QUERY_KIND.LineName,
        line: 10,
      });
    }
  });

  test("parses comma-separated tokens", () => {
    const r = parseRootQueries("10:foo,42,9-13:bar");
    expect(r.ok).toEqual(true);
    if (r.ok) {
      expect(r.queries).toHaveLength(3);
      expect(r.queries.map((v) => v.kind)).toEqual([
        "line-name",
        "line",
        "range-name",
      ]);
    }
  });

  test("parses L-prefixed forms in a comma list", () => {
    const r = parseRootQueries("L10,L5-10,L20");
    expect(r.ok).toEqual(true);
    if (r.ok) {
      expect(r.queries).toHaveLength(3);
      expect(r.queries[0]).toMatchObject({
        kind: ROOT_QUERY_KIND.LineOrName,
        line: 10,
        name: "L10",
      });
      expect(r.queries[1]).toMatchObject({
        kind: ROOT_QUERY_KIND.Range,
        start: 5,
        end: 10,
        raw: "L5-10",
      });
      expect(r.queries[2]).toMatchObject({
        kind: ROOT_QUERY_KIND.LineOrName,
        line: 20,
        name: "L20",
      });
    }
  });

  test("rejects empty value", () => {
    expect(parseRootQueries("")).toEqual({
      ok: false,
      error: "empty --roots value",
    });
  });

  test("rejects values with whitespace around tokens", () => {
    expect(parseRootQueries("10, 42")).toMatchObject({ ok: false });
    expect(parseRootQueries(" 10")).toMatchObject({ ok: false });
  });

  test("rejects trailing comma", () => {
    expect(parseRootQueries("10,")).toMatchObject({ ok: false });
    expect(parseRootQueries(",10")).toMatchObject({ ok: false });
  });

  test("rejects an empty token in the middle of a comma list", () => {
    expect(parseRootQueries("10,,42")).toMatchObject({ ok: false });
  });

  test("propagates the offending token in the error", () => {
    const r = parseRootQueries("10,foo-bar");
    expect(r.ok).toEqual(false);
    if (!r.ok) {
      expect(r.error).toContain("foo-bar");
    }
  });
});

File: crates/unsnarl-root-query/src/parse_root_queries_test.rs
use super::*;
use unsnarl_ir::SourceLine;

#[test]
fn parses_single_token() {
    let queries = parse_root_queries("10:foo").expect("should parse");
    assert_eq!(queries.len(), 1);
    assert!(matches!(
        queries[0],
        ParsedRootQuery::LineName {
            line: SourceLine(10),
            ..
        },
    ));
}

#[test]
fn parses_comma_separated_tokens() {
    let queries = parse_root_queries("10:foo,42,9-13:bar").expect("should parse");
    assert_eq!(queries.len(), 3);
    assert!(matches!(queries[0], ParsedRootQuery::LineName { .. }));
    assert!(matches!(queries[1], ParsedRootQuery::Line { .. }));
    assert!(matches!(queries[2], ParsedRootQuery::RangeName { .. }));
}

#[test]
fn parses_l_prefixed_forms_in_comma_list() {
    let queries = parse_root_queries("L10,L5-10,L20").expect("should parse");
    assert_eq!(queries.len(), 3);
    assert_eq!(
        queries[0],
        ParsedRootQuery::LineOrName {
            line: SourceLine(10),
            name: "L10".to_string(),
            raw: "L10".to_string(),
        },
    );
    assert_eq!(
        queries[1],
        ParsedRootQuery::Range {
            start: SourceLine(5),
            end: SourceLine(10),
            raw: "L5-10".to_string(),
        },
    );
    assert_eq!(
        queries[2],
        ParsedRootQuery::LineOrName {
            line: SourceLine(20),
            name: "L20".to_string(),
            raw: "L20".to_string(),
        },
    );
}

#[test]
fn rejects_empty_value() {
    assert_eq!(
        parse_root_queries(""),
        Err("empty --roots value".to_string())
    );
}

#[test]
fn rejects_whitespace_around_tokens() {
    assert!(parse_root_queries("10, 42").is_err());
    assert!(parse_root_queries(" 10").is_err());
}

#[test]
fn rejects_trailing_or_leading_comma() {
    assert!(parse_root_queries("10,").is_err());
    assert!(parse_root_queries(",10").is_err());
}

#[test]
fn rejects_empty_token_in_middle() {
    assert!(parse_root_queries("10,,42").is_err());
}

#[test]
fn propagates_offending_token_in_error() {
    let err = parse_root_queries("10,foo-bar").unwrap_err();
    assert!(
        err.contains("foo-bar"),
        "expected error to mention 'foo-bar', got {err:?}",
    );
}
